enhancements/network/ovn-kubernetes-ipsec.md

```diff
@@ -0,0 +1,298 @@
+identify them early and start the process.
+
+#### Performance
+
+Enablement of IPsec will cause degradation of overall networking throughput and
+latency due to the additional overhead of the IPsec protocol. Some early
+performance testing will be done in a non-production cluster in order to get
+an approximation of the expected performance degradation - These results will be
+posted at a later date. After implementation of the enhancement, it is expected
+that comprehensive benchmarking and profiling will take place. Based on the
+output of this activity, it may be decided to start some optimization activities
+in parallel to improve performance.
+
+#### Security Review
+
+See "Open Questions" below.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Will this need to go through some kind of Security Review?
+
+### Test Plan
+
+It will expected to add a number of Continuous Integration jobs for IPsec:
+
+- All Openshift e2e tests should be replicated in a cluster that has been
+  installed with IPsec enabled. This should ensure IPsec has no regressions with
+  respect to a cluster that has not been enabled with IPsec.
+- In order to test any issues with version skew, an IPsec-enabled cluster should
+  be upgraded while active.
+
+Additional CI jobs could potentially be added at a later stage.
+
+- Functional tests
+    - Ensure certificate expiration and rotation happens correctly.
+- Non-functional tests
+    - Performance tests (scale tests)
+    - Endurance tests (traffic generation over a long period)
+    - Recovery tests (what happens when node IPsec components such as pluto or
+      ovs-monitor-ipsec are killed)
+
+It should also be noted that, some upstream OVS/OVN unit and integration tests
+will also need to be developed.
+
+### Graduation Criteria
+
+TBD
+
+#### Examples
+
+##### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Performance measurement
+
+##### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+##### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+Currently, it is proposed that this feature can only be enabled at cluster
+installation time.
+
+### Version Skew Strategy
+
+It is proposed that this enhancement will not support mixed clusters of RHEL7
+and RHEL8 nodes. This can be supported in a later enhancement.
+
+However, it will need to support version skew within RHEL8 components. The exact
+details of this will be understood, when it is determined how IPsec components
+will be containerized but some considerations are noted below:
+
+- Internode libreswan compatibility
+  - Each nodes' libreswan version will need to be compatible with other nodes'
+    libreswan version
```

Comment:
How so? What sort of changes could there be that would break us?

Reply:
I am not an expert on libreswan, but I am assuming incompatibilities in IKE daemons could cause issues. However, I am at a loss to think of an example (especially because, I assume, libreswan is based off the IKE standard). I think I can remove this.

Reply:
Your best bet would be to ensure the same Libreswan version on all nodes. It can interoperate with Strongswan and other IKEv2 implementations but sometimes only in certain configurations.

Reply:
@mrogers950 if there were version skews, this could cause problems?

Reply:
I'm going to leave this in based on @mrogers950 feedback.